### DISCUSSION TITLE
Overleaf编译与word效果差异，且无法设置fontset

### DISCUSSION THREAD

**Discussion by QihaoWan** [tuna/thuthesis#747] [Q&A] [answered]
使用Overleaf在Windows编译获得pdf时，正文字体看起来比word输出的字体略显偏大，在手动设置\documentclass[degree=doctor,fontset=windows]{thuthesis}后，则编译失败，无法获得编译结果，问要怎么调整使得输出字体与word一致

**Comment by Harry-Chen** [comment, marked as answer]
请在 Windows 本地编译，或者上传 Windows 相应字体。受限于版权原因，模板无法提供这些字体。

**Reply by QihaoWan** [reply]
可以请教如何上传Windows相应字体吗

**Reply by Harry-Chen** [reply]
需要将 `SimSun`, `SimHei`, `FangSong`, `KaiTi`, `LiSu` 等字体从 Windows 字体目录上传到 Overleaf

**Reply by zepinglee** [reply]
宋黑仿楷就可以，不需要隶书 LiSu。

**Reply by LUKE-ZZ** [reply]
😂作为小白我有点整不明白，我在windows字库里找到了 SimSun, SimHei, FangSong, KaiTi, LiSu等字体，上传到了overleaf根目录下，这样可以吗？怎么让overleaf知道我要用这些字体呢？

**Reply by LUKE-ZZ** [reply]
已通过其他方式解决，还是谢谢啦

**Reply by zepinglee** [reply]
> 😂作为小白我有点整不明白，我在windows字库里找到了 SimSun, SimHei, FangSong, KaiTi, LiSu等字体，上传到了overleaf根目录下，这样可以吗？怎么让overleaf知道我要用这些字体呢？

`\documentclass[..., fontset = windows]{thuthesis}`

**Reply by LUKE-ZZ** [reply]
> > 😂作为小白我有点整不明白，我在windows字库里找到了 SimSun, SimHei, FangSong, KaiTi, LiSu等字体，上传到了overleaf根目录下，这样可以吗？怎么让overleaf知道我要用这些字体呢？
> 
> `\documentclass[..., fontset = windows]{thuthesis}`

https://www.overleaf.com/3596431325dhghdntydmmd

不好意思，我已经上传了 SimSun, SimHei, FangSong, KaiTi, LiSu等字体，也加入了\documentclass[..., fontset = windows]{thuthesis}这个命令，可以还是会报错，请问有没有可能可以看一下问题在哪里呢，下面是我尝试对模板中修改字体的文档。谢谢！

https://www.overleaf.com/3596431325dhghdntydmmd

**Reply by LUKE-ZZ** [reply]
我尝试了\setCJKsansfont{}但是也没效果（我乱试的，因为我是真的小白😂）

**Reply by zepinglee** [reply]
> 我尝试了\setCJKsansfont{}但是也没效果（我乱试的，因为我是真的小白😂）

简单看了一下，似乎不能按照字体名调用当前目录下的字体，需要修改 .cls 文件中 `\thu@set@cjk@font@windows` 的内容，将字体名改为文件名。确实比较麻烦，但是似乎没有更好的办法。

**Comment by LUKE-ZZ** [comment]
OverLeaf中没有和学校标准中完全一致的标准的宋体黑体，答辩秘书老师看了我论文，说字体很奇怪

尝试上传字体发现总是有问题。也尝试了使用一些命令，但是效果都不理想。找到一个解决方法如下：

首先使用一个windows电脑安装liveTEX
http://mirrors.tuna.tsinghua.edu.cn/CTAN/systems/texlive/tlnet/install-tl-windows.exe
时间可能会比较久，我花了3小时。

安装完成后abort按钮会变灰，有welcome界面

在OverLeaf网站上打开文档后，点击左上角的OverLeaf叶子图标，点击"Source"下载整个文档的压缩包，解压，找到tex主文件，会默认用texwork editor打开，点击绿色编译按钮即可得到非常标准字体的文件

**Reply by LUKE-ZZ** [reply]
而且我发现不知道为什么Overleaf下载下来的包在texworks editor里编译，需要XeLatex编译两次，BiTex编译一次，XeLatex编译一次，BiTex编译一次，XeLatex编译一次，这样吊诡的顺序之后，才会交叉引用和参考文献都显示正常

**Comment by caizefeng** [comment]
To compile `ThuThesis (v7.3.2)` on **overleaf** using `fontset=windows`, you can **upload** four font files `simfang.ttf`, `simhei.ttf`, `simkai.ttf`, `simsun.ttc` to the overleaf project directory, and **modify** two blocks, `\ifthu@cjk@font@auto` and `\thu@set@cjk@font@windows`, in `thuthesis.cls` as 

```latex
\ifthu@cjk@font@auto
  \ifthu@fontset@mac
    \thusetup{cjk-font = mac}
  \else
    \ifthu@fontset@windows
      \thusetup{cjk-font = windows}
    \else
      \ifthu@fontset@ubuntu
        \thusetup{cjk-font = noto}
      \else
        \thusetup{cjk-font = fandol}
      \fi
    \fi
  \fi
\fi
```

```latex
\newcommand\thu@set@cjk@font@windows{%
  \setCJKmainfont{simsun}[
    Path          = ./,
    Extension     = .ttc,
    AutoFakeBold  = 3,
    ItalicFont    = simkai,
    ItalicFeatures = {Path = ./, Extension = .ttf},
  ]%
  \setCJKsansfont{simhei}[
    Path          = ./,
    Extension     = .ttf,
    AutoFakeBold  = 3,
  ]%
  \setCJKmonofont{simfang}[
    Path          = ./,
    Extension     = .ttf,
  ]%
  \setCJKfamilyfont{zhsong}{simsun}[
    Path          = ./,
    Extension     = .ttc,
    AutoFakeBold  = 3,
  ]%
  \setCJKfamilyfont{zhhei}{simhei}[
    Path          = ./,
    Extension     = .ttf,
    AutoFakeBold  = 3,
  ]%
  \setCJKfamilyfont{zhkai}{simkai}[
    Path          = ./,
    Extension     = .ttf,
  ]%
  \setCJKfamilyfont{zhfs}{simfang}[
    Path          = ./,
    Extension     = .ttf,
  ]%
}
```
Also, change the `fontset` argument to `windows`. Note that the generated PDF will be **visually identical** to the one generated on Windows platform but with slightly different size due to varying metadata. The reason you don't need to upload other Windows fonts like `Times New Roman` and `Arial` is because Overleaf [pre-installs](https://www.overleaf.com/learn/latex/XeLaTeX) them.

**Reply by LUKE-ZZ** [reply]
Thank you so much!~

**Comment by zepinglee** [comment]
可以考虑使用 TeXPage，其中提供了 Windows 中文字体。

**Reply by thushaoyu** [reply]
请问texpage编译出来只有一页，是什么原因呢？

**Reply by zepinglee** [reply]
> 请问texpage编译出来只有一页，是什么原因呢？

新开 issue，给出复现方式。

**Comment by zepinglee** [comment]
eb938d8 修改为，若系统不存在 SimSun 字体，会尝试调用 Windows 字体的文件（`Simsun.ttc`, `SimHei.ttf`, `Kaiti.ttf`, `Fangsong.ttf`）。默认从当前目录寻找，也可以在 `\documentclass` 参数中使用 `windows-font-dir` 设置字体的目录，例如字体在 `fonts/` 目录下，则设置 `\documentclass[degree=master, windows-font-dir=fonts]{thuthesis}`。

**Reply by Harry-Chen** [reply]
```tex
  \setCJKmainfont{SimSun}[%
    Path         = \thu@windows@font@dir/,
    AutoFakeBold = 3,
    ItalicFont   = KaiTi,
  ]%
```

这样能够在对应的路径下找到 posix 字体名吗？是否需要指定到具体的字体文件路径？

**Reply by zepinglee** [reply]
这个是按照文件名（忽略大小写）找的，不是字体名。

**Reply by Harry-Chen** [reply]
Windows 的中易楷体名字叫 simkai.ttf

**Reply by zepinglee** [reply]
这……<https://learn.microsoft.com/en-us/typography/font-list/kaiti> 确实。我是按照 Mac 版 Word 字体的文件名写的。

**Reply by zepinglee** [reply]
> Windows 的中易楷体名字叫 simkai.ttf

在 17bd760 中解决。

**Reply by jjyyxx** [reply]
如果是在本地 Linux 机器中部署的 sharelatex 容器，里面没有预装 `Times New Roman` 和 `Arial`，能否像 `windows-font-dir` 一样处理呢？（目前我是在有字体需求时，传到Overleaf上处理的）

**Reply by zepinglee** [reply]
应该可以……

感觉这模板 40% 的代码在配置字体。

**Reply by jjyyxx** [reply]
Windows的Arial和Times New Roman复制出来长这样
```
arial.ttf
arialbd.ttf
arialbi.ttf
ariali.ttf
ARIALN.TTF
ARIALNB.TTF
ARIALNBI.TTF
ARIALNI.TTF
ariblk.ttf
times.ttf
timesbd.ttf
timesbi.ttf
timesi.ttf
```
然后我自己试了一下
```
\documentclass{article}
\usepackage{fontspec}
\setmainfont{Times New Roman}[Path = ./]
\title{}\author{}
\begin{document}A\end{document}
```
就找不到，把`Times New Roman`改成`times`倒是可以，但斜体之类就不会配置了，`Arial`看起来就更复杂了。而Overleaf和Windows上把`[Path = ./]`去掉确实是可用的。

对我自己来说，因为容器是我自己部署的，我也可以再mount一些字体进去，不一定需要改模板，但是直接把上面这些`.ttf`放进去我觉得应该一样不行，不知道是不能用这几个字体文件还是`setmainfont`要改，又或者是其他环节的问题。希望能指导一下。

**Reply by jjyyxx** [reply]
我试了一下，可以直接给容器装这几个英文字体，装好就没问题了。

Dockerfile中的新增部分供参考：

```Dockerfile
RUN \
    echo ttf-mscorefonts-installer msttcorefonts/accepted-mscorefonts-eula select true | debconf-set-selections && \
    apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y ttf-mscorefonts-installer && rm -rf /var/lib/apt/lists/*
```

如果一并装中文字体的话
```Dockerfile
# > ls sim
# simfang.ttf  simhei.ttf  simkai.ttf  simsun.ttc
COPY sim /usr/share/fonts/truetype/sim
RUN fc-cache -f -v
```
之后就可以直接用`fontset=windows`。

**Reply by muzimuzhi** [reply]
> [eb938d8](https://github.com/tuna/thuthesis/commit/eb938d8ebc87ee68270ab7bdd3d281b6db5524fa) 修改为，若系统不存在 SimSun 字体，会尝试调用 Windows 字体的文件……

@zepinglee 这里是不是描述错了，是「若系统存在 SimSun 字体」，而不是「若系统**不**存在 SimSun」。
